app/src/dex/router/fill_route2.rs

```diff
@@ -0,0 +1,370 @@
+            input.amount = input.amount - current_input.amount;
+            output.amount = output.amount + current_output.amount;
+
+            tracing::debug!(
+                current_input = ?current_input.amount,
+                current_output = ?current_output.amount,
+                input = ?input.amount,
+                output = ?output.amount,
+                "completed fill iteration"
+            );
+
+            let current_effective_price =
+                U128x128::from(current_input.amount) / U128x128::from(current_output.amount);
+
+            // Check whether we should stop filling because we've exceeded the spill price:
+            // TODO: is comparing Options the right behavior here? can we
+            // continue rolling over dust positions without allowing arbitrarily
+            // higher prices than the requested spill price?
+            if current_effective_price > spill_price {
+                tracing::debug!(
+                    // HACK: render infinite price as 0
+                    spill_price = %spill_price.unwrap_or_default(),
+                    current_effective_price = %current_effective_price.unwrap_or_default(),
+                    "exceeded spill price, breaking loop"
+                );
+                break 'filling;
+            }
+
+            // Find the next position to replace the one we just consumed.
+            'next_position: loop {
+                let next_position_id = match positions_by_price
+                    .get_mut(&pairs[constraining_index])
+                    .unwrap()
+                    .as_mut()
+                    .next()
+                    .await
+                    .transpose()?
+                {
+                    // If none is available, we can't keep filling, and need to break.
+                    None => {
+                        tracing::debug!(
+                            pair = ?pairs[constraining_index],
+                            "no more positions available for pair, breaking loop"
+                        );
+                        break 'filling;
+                    }
+                    // Otherwise, we need to check that the position is not already
+                    // part of the current frontier.
+                    Some(position_id) if !frontier_position_ids.contains(&position_id) => {
+                        position_id
+                    }
+                    // Otherwise, continue to the next position in the stream.
+                    Some(_) => continue 'next_position,
+                };
+
+                let next_position = self
+                    .position_by_id(&next_position_id)
+                    .await?
+                    .expect("indexed position should exist");
+
+                frontier_position_ids.insert(next_position_id);
+                frontier[constraining_index] = next_position;
+                // TODO: should we remove the old position's ID from the ID set?
```

Comment:
I think, but I'm not certain, that we are missing a `break 'next_position`. AFAICT, it would make sense to have it here when we have a bona fide `next_position` to roll with